source/common/http/filter_manager.h

```diff
@@ -857,6 +863,16 @@ class FilterManager : public ScopeTrackedObject,
 
   virtual bool shouldLoadShed() { return false; };
 
+  void sendGoAwayAndClose() {
+    // Stop filter chain iteration by checking encoder or decoder chain.
+    if (state_.filter_call_state_ & FilterCallState::IsDecodingMask) {
+      state_.decoder_filter_chain_aborted_ = true;
+    } else if (state_.filter_call_state_ & FilterCallState::IsEncodingMask) {
```

Comment:
This can potentially be a problem, if a decoder filter calls `sendLocalReply()` and an encoder filter processing local reply calls `sendGoAwayAndClose()`. It is possible that encoder filter chain my not be aborted in this case.

Can you construct a test case that validates this, please?

Reply:
Good point, thanks Yan.

I examined the code more, and the sate is flipped back through `state_.filter_call_state_ &= ~FilterCallState::DecodeHeaders;` after each filter's execution, so this will not be true in encoder cases, and if it is sending local reply in decoder filter chain, it is deferred [here](https://github.com/envoyproxy/envoy/blob/4aa1d1b2b7bc408d5b61ee2eb389e31300b76ad0/source/common/http/filter_manager.cc#L1007), so the mask is always reset back. Added a test for it.